internal/testoutput/disentangle.go

```diff
@@ -0,0 +1,77 @@
+package testoutput
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+// TestEvent taken from `go doc test2json`, unfortunately, this struct is not
+// exposed from any core go packages so we have to define it ourselves
+type TestEvent struct {
+	Time    time.Time `json:",omitempty"`
+	Action  string
+	Package string  `json:",omitempty"`
+	Test    string  `json:",omitempty"`
+	Elapsed float64 `json:",omitempty"`
+	Output  string  `json:",omitempty"`
+}
+
+// DisentangleTestOutputs separates test outputs from a stream of interleaved
+// test events into separate files per individual testcase
+func DisentangleTestOutputs(testResultsPath string, in *os.File, out *os.File) error {
+	// holds a mapping from test to the file handles the output is redirected to
+	var fileHandles = make(map[string]*os.File)
+	decoder := json.NewDecoder(in)
+
+	for {
+		var event TestEvent
+		if err := decoder.Decode(&event); err == io.EOF {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		// switch over fixed set of actions (c.f. `go doc test2json`)
+		switch event.Action {
+		case "run":
+			filePath := filepath.Join(testResultsPath, event.Test+".log")
+			dir := filepath.Dir(filePath)
+			err := os.MkdirAll(dir, 0777)
+			if err != nil {
+				return fmt.Errorf("failed to create directory: %w", err)
+			}
+
+			f, err := os.OpenFile(filePath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
```

Comment:
I am curious did you need this file flags and permissions? Have you different flags/permissions first - perhaps [os.Create()](https://cs.opensource.google/go/go/+/refs/tags/go1.17.6:src/os/file.go;l=326) and it did not work?

Reply:
Good catch, we'll check `os.Create()` instead, looks like this is a convenience wrapper around `os.OpenFile()`